book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+heights to blocks. Chains are stored using an ordered map from difficulty to
+chains, so that the map ordering is the ordering of best to worst chains.
+
+### The `Chain` type
+[chain-type]: #chain-type
+
+
+The `Chain` type represents a chain of blocks. Each block represents an
+incremental state update, and the `Chain` type caches the cumulative state
+update from its root to its tip.
+
+The `Chain` type is used to represent the non-finalized portion of a complete
+chain of blocks rooted at the genesis block. The parent block of the root of
+a `Chain` is the tip of the finalized portion of the chain.
+
+The `Chain` type supports serveral operations to manipulate chains, `push`,
+`pop_root`, and `fork`. `push` is the most fundamental operation and handles
+contextual validation of chains as they are extended. `pop_root` is provided
+for finalization, and is how we move blocks from the non-finalized portion of
+the state to the finalized portion. `fork` on the other hand handles creating
+new chains for `push` when new blocks arrive whose parent isn't a tip of an
+existing chain.
+
+The `Chain` type is defined by the following struct and API:
+
+```rust
+struct Chain {
+    blocks: BTreeMap<block::Height, Arc<Block>>,
+    height_by_hash: HashMap<block::Hash, block::Height>,
+    tx_by_hash: HashMap<transaction::Hash, (block::Height, tx_index)>,
+
+    utxos: HashSet<transparent::Output>,
+    sapling_anchors: HashSet<sapling::tree::Root>,
+    sprout_anchors: HashSet<sprout::tree::Root>,
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+    partial_cumulative_work: PartialCumulativeWork,
+}
+```
+
+#### `pub fn push(&mut self, block: Arc<Block>) -> Result<(), Error>`
+
+Push a block into a chain as the new tip if the block is a valid extension of
+that chain.
+
+1. Run contextual validation checks on block against Self
+
+1. Update cumulative data members
+    - Add block to end of `self.blocks`
+    - Add hash to `height_by_hash`
+    - for each `transaction` in `block`
+      - add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+    - Add new utxos and remove consumed utxos from `self.utxos`
+    - Add anchors to the appropriate `self.<version>_anchors`
+    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add work to `self.partial_cumulative_work`
+
+#### `pub fn pop_root(&mut self) -> Arc<Block>`
+
+Remove the lowest height block of the non-finalized portion of a chain.
+
+1. Remove the lowest height block from `self.blocks`
+
+1. Update cumulative data members
+    - Remove the block's hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Remove new utxos from `self.utxos`
+    - Remove the anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
```

Comment:
```suggestion
1. Update block header cumulative data members
    - Remove the block's hash from `self.height_by_hash`
    
1. If the block is **not** a genesis block, update transaction cumulative data members
    - for each `transaction` in `block`
      - remove `transaction.hash` from `tx_by_hash`
    - Remove new utxos from `self.utxos`
    - Remove the anchors from the appropriate `self.<version>_anchors`
    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
```

Reply:
same as above here, it looks like this is fine as is.